0000-conditional_dependencies.md

```diff
@@ -0,0 +1,142 @@
+turn on a dependency. For example, if I have a library which defines a new
+type, I might want users to be able to serialize it with `serde`, but I don't
+want users who don't care about this functionality to have to add `serde` to
+their dependency graph. So I add a `serialization` feature to my Cargo.toml,
+and only if you turn on this feature do you get the feature for serialization.
+
+Ultimately, the source of this problem is the orphan rules - because only *I*
+can define the impls for my type, I have to add this dependency. For every
+commonly used trait defining library, I have to add a new feature, and all of
+my clients have to turn on that feature.
+
+cargo features have many uses, but this particular pattern ends up being a lot
+of fiddling for a very common use case. Instead of requiring users go through
+the feature flag rigmarole, we can provide first class support for this
+pattern through "conditional dependencies" - dependencies that are turned on
+automatically if they were already present.
+
+# Detailed design
+[design]: #detailed-design
+
+## Conditional dependencies in `Cargo.toml`
+
+When declaring a dependency in a Cargo.toml file, if that dependency is an
+object, it can contain the `conditional` member, which is a boolean, similar to
+the `optional` dependency. For example:
+
+```toml
+[dependencies]
+foobar = { version = "1.0.0", conditional = true }
+```
+
+Or:
+
+```toml
+[dependencies.foobar]
+version = "1.0.0"
+conditional = true
+```
+
+A single dependency objecy cannot contain both a `conditional` key and an
+`optional` key.
+
+When generating the dependency graph, cargo will not include dependencies
+tagged conditional. It will then traverse the graph looking for conditional
+dependencies; if a matching source for a conditional dependency (e.g. a
+matching version number) is already present in the graph, cargo will
+automatically add that dependency.
+
+## `#[cfg(dependency=)]`
+
+A new kind of cfg attribute will be added to rustc. The `dependency` attribute
+will only be compiled when a certain dependency has been explicitly passed to
+rustc using the --extern flag.
+
+Because cargo automatically passes dependencies explicitly with this command,
+this code will be compiled only when cargo triggers the conditional dependency.
+
+## Convention for using conditional dependencies
+
+The above to features can be composed in any way, but to assist in adopting
+this functionality, we can define an encouraged convention.
+
+If you want to add a new conditional dependency, you should add a top level
+submodule (that is, mounted directly under lib, main, etc) with the same name
+as the conditional dependency. This module will be tagged with the cfg, and the
+extern crate declaration and all code that uses symbols from that crate will
+go inside of that module.
+
+For example:
+
+```rust
+/// lib.rs
+
+#[cfg(dependency=foobar]
+mod foobar;
+```
+
+```rust
+/// foobar.rs
+
+extern crate foobar;
```

Comment:
Note that this is a departure from the *very* strong convention of only using `extern crate` in the crate root. It also has substantial potential to be confusing, because now for to import something from that crate you’d need `use self::foobar::Bar;` or `use foobar::foobar::Bar;` rather than `use foobar::Bar;`.